gems/aws-eventstream/lib/aws-eventstream/decoder.rb

```diff
@@ -0,0 +1,247 @@
+require 'stringio'
+require 'tempfile'
+require 'zlib'
+
+module Aws
+  module EventStream 
+
+    # This class provides method for decoding binary inputs into
+    # single or multiple messages (Aws::EventStream::Message).
+    #
+    # * {#decode} - decodes messages from an IO like object responds
+    #   to #read that containing binary data, returning decoded
+    #   Aws::EventStream::Message along the way or wrapped in a enumerator
+    #
+    # ## Examples
+    #
+    #   decoder = Aws::EventStream::Decoder.new
+    #
+    #   # decoding from IO
+    #   decoder.decode(io) do |message|
+    #     message.headers
+    #     # => { ... }
+    #     message.payload
+    #     # => StringIO / Tempfile
+    #   end
+    #
+    #   # alternatively
+    #   message_pool = decoder.decode(io)
+    #   message_pool.next
+    #   # => Aws::EventStream::Message
+    #   
+    # * {#decode_chunk} - decodes a single message from a chunk of data,
+    #   returning message object followed by boolean(indicating eof status
+    #   of data) in an array object
+    #
+    # ## Examples
+    #
+    #   # chunk containing exact one message data
+    #   message, chunk_eof = decoder.decode_chunk(chunk_str)
+    #   message
+    #   # => Aws::EventStream::Message
+    #   chunk_eof
+    #   # => true
+    #
+    #   # chunk containing partial of single message data
```

Comment:
"chunk containing a partial message" is what I think we're trying to say here, this one is more unclear